unused_parens false positive on addition of blocks?

In this example:

```rust
pub fn foo(a: bool, b: bool) -> u8 {
    (if a { 1 } else { 0 } + if b { 1 } else { 0 })
}
```

rustc throws a warning:
```
warning: unnecessary parentheses around block return value
 --> src/lib.rs:2:5
  |
2 |     (if a { 1 } else { 0 } + if b { 1 } else { 0 })
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove these parentheses
  |
  = note: `#[warn(unused_parens)]` on by default
```

Removing the parentheses as the message directs causes errors:

```
error: expected expression, found `+`
 --> src/lib.rs:2:27
  |
2 |     if a { 1 } else { 0 } + if b { 1 } else { 0 }
  |                           ^ expected expression

error[E0308]: mismatched types
 --> src/lib.rs:2:12
  |
2 |     if a { 1 } else { 0 } + if b { 1 } else { 0 }
  |     -------^-------------- help: consider using a semicolon here
  |     |      |
  |     |      expected `()`, found integer
  |     expected this to be `()`

error[E0308]: mismatched types
 --> src/lib.rs:2:23
  |
2 |     if a { 1 } else { 0 } + if b { 1 } else { 0 }
  |     ------------------^--- help: consider using a semicolon here
  |     |                 |
  |     |                 expected `()`, found integer
  |     expected this to be `()`

error: aborting due to 3 previous errors
```

It seems the warning is wrong, unless this is expected to compile

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.42.0 (b8cedc004 2020-03-09)
binary: rustc
commit-hash: b8cedc00407a4c56a3bda1ed605c6fc166655447
commit-date: 2020-03-09
host: x86_64-unknown-linux-gnu
release: 1.42.0
LLVM version: 9.0
```

This also happens on nightly.
```
rustc 1.44.0-nightly (ce93331e2 2020-04-17)
binary: rustc
commit-hash: ce93331e2cf21ac4b72a53854b105955919114e7
commit-date: 2020-04-17
host: x86_64-unknown-linux-gnu
release: 1.44.0-nightly
LLVM version: 9.0
```


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->
This issue has been assigned to @mibac138 via [this comment](https://github.com/rust-lang/rust/issues/71290#issuecomment-623767470).
<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"mibac138"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

I think this may apply to any binary op where the left-hand side is an expression that can be a statement without a semicolon, per:
https://github.com/rust-lang/rust/blob/45d050cde277b22a755847338f2acc2c7b834141/src/librustc_ast/util/classify.rs#L14-L25

For example, adding with block expressions also reports unused parens:

```rust
pub fn foo(a: bool, b: bool) -> u8 {
    ({ u8::from(a) } + { u8::from(b) })
}
```

```
warning: unnecessary parentheses around block return value
 --> src/lib.rs:2:5
  |
2 |     ({ u8::from(a) } + { u8::from(b) })
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove these parentheses
  |
  = note: `#[warn(unused_parens)]` on by default
```

but removing those gets an error:

```
error: expected expression, found `+`
 --> src/lib.rs:2:21
  |
2 |     { u8::from(a) } + { u8::from(b) }
  |     --------------- ^ expected expression
  |     |
  |     help: parentheses are required to parse this as an expression: `({ u8::from(a) })`

error[E0308]: mismatched types
 --> src/lib.rs:2:7
  |
2 |     { u8::from(a) } + { u8::from(b) }
  |       ^^^^^^^^^^^- help: try adding a semicolon: `;`
  |       |
  |       expected `()`, found `u8`

error: aborting due to 2 previous errors
```

As a workaround, it doesn't trigger `unused_parens` if the operands are wrapped separately:

```rust
pub fn foo(a: bool, b: bool) -> u8 {
    (if a { 1 } else { 0 }) + (if b { 1 } else { 0 })
}
```

or

```rust
pub fn foo(a: bool, b: bool) -> u8 {
    ({ u8::from(a) }) + ({ u8::from(b) })
}
```

--- a/src/librustc_lint/unused.rs
+++ b/src/librustc_lint/unused.rs
@@ -380,11 +380,19 @@ trait UnusedDelimLint {
     );
 
     fn is_expr_delims_necessary(inner: &ast::Expr, followed_by_block: bool) -> bool {
-        followed_by_block
-            && match inner.kind {
-                ExprKind::Ret(_) | ExprKind::Break(..) => true,
-                _ => parser::contains_exterior_struct_lit(&inner),
+        // Prevent false-positives in cases like `fn x() -> u8 { ({ 0 } + 1) }`
+        let lhs_needs_parens = match &inner.kind {
+            ExprKind::Binary(_, lhs, _rhs) => {
+                !rustc_ast::util::classify::expr_requires_semi_to_be_stmt(&*lhs)
             }
+            _ => false,
+        };
+        lhs_needs_parens
+            || (followed_by_block
+                && match inner.kind {
+                    ExprKind::Ret(_) | ExprKind::Break(..) => true,
+                    _ => parser::contains_exterior_struct_lit(&inner),
+                })
     }
 
     fn emit_unused_delims_expr(

--- a/src/librustc_lint/unused.rs
+++ b/src/librustc_lint/unused.rs
@@ -381,12 +381,20 @@ trait UnusedDelimLint {
 
     fn is_expr_delims_necessary(inner: &ast::Expr, followed_by_block: bool) -> bool {
         // Prevent false-positives in cases like `fn x() -> u8 { ({ 0 } + 1) }`
-        let lhs_needs_parens = match &inner.kind {
-            ExprKind::Binary(_, lhs, _rhs) => {
-                !rustc_ast::util::classify::expr_requires_semi_to_be_stmt(&*lhs)
+        let lhs_needs_parens = {
+            let mut innermost = inner;
+            loop {
+                if let ExprKind::Binary(_, lhs, _rhs) = &innermost.kind {
+                    innermost = lhs;
+                    if !rustc_ast::util::classify::expr_requires_semi_to_be_stmt(innermost) {
+                        break true;
+                    }
+                } else {
+                    break false;
+                }
             }
-            _ => false,
         };
+
         lhs_needs_parens
             || (followed_by_block
                 && match inner.kind {

--- a/src/test/ui/lint/issue-71290-unused-paren-binop.rs
+++ b/src/test/ui/lint/issue-71290-unused-paren-binop.rs
@@ -1,0 +1,17 @@
+// check-pass
+// Make sure unused parens lint doesn't emit a false positive.
+// See https://github.com/rust-lang/rust/issues/71290 for details.
+
+fn x() -> u8 {
+    ({ 0 }) + 1
+}
+
+fn y() -> u8 {
+    ({ 0 } + 1)
+}
+
+pub fn foo(a: bool, b: bool) -> u8 {
+    (if a { 1 } else { 0 } + if b { 1 } else { 0 })
+}
+
+fn main() {}

--- a/src/test/ui/lint/issue-71290-unused-paren-binop.rs
+++ b/src/test/ui/lint/issue-71290-unused-paren-binop.rs
@@ -1,6 +1,7 @@
 // check-pass
 // Make sure unused parens lint doesn't emit a false positive.
 // See https://github.com/rust-lang/rust/issues/71290 for details.
+#![deny(unused_parens)]
 
 fn x() -> u8 {
     ({ 0 }) + 1

--- a/src/test/ui/lint/issue-71290-unused-paren-binop.rs
+++ b/src/test/ui/lint/issue-71290-unused-paren-binop.rs
@@ -15,4 +15,9 @@ pub fn foo(a: bool, b: bool) -> u8 {
     (if a { 1 } else { 0 } + if b { 1 } else { 0 })
 }
 
+pub fn bar() -> u8 {
+    // Make sure nested expressions are handled correctly as well
+    ({ 0 } + 1 + 2 + 3 + 4 + 5 + 6 + 7 + 8 + 9)
+}
+
 fn main() {}

